### DISCUSSION TITLE
[FIX] podman lxc is working on zfs with this fix

### DISCUSSION THREAD

**Discussion by tctlrd** [tteck/Proxmox#3531] [Show and tell]
### Please verify that you have read and understood the guidelines.

yes

### A clear and concise description of the issue.

There is a workaround for making podman lxc run on zfs:
https://medium.com/@rar1871/docker-podman-in-lxc-with-zfs-5196a13bb8be

### What settings are you currently utilizing?

Default Settings

### Which Linux distribution are you employing?

Debian 12

### Please provide detailed steps to reproduce the issue.

> Create a mount wrapper.
> 
> nano /usr/local/bin/overlayzfsmount
> 
> Paste the following code into the file:
> 
> #!/bin/sh
> exec /bin/mount -t overlay overlay "$@"
> 
> Save and close the file. Then make it executable by running
> 
> chmod +x /usr/local/bin/overlayzfsmount
> 
> 2. Edit Podman’s storage configuration. If the file does not exist, go ahead and create it.
> 
> nano /etc/containers/storage.conf
> 
> Note: If you are running in rootless mode the path to this file will be different. It will most likely be $HOME/.config/containers/storage.conf
> 
> Add the following to the file
> 
> [storage]
> driver = "overlay"
> runroot = "/run/containers/storage"
> graphroot = "/var/lib/containers/storage"
> 
> [storage.options]
> mount_program = "/usr/local/bin/overlayzfsmount"

**Comment by alexhaydock** [comment]
This works quite well for _running_ containers, but I think there's still some issues using Podman in LXC on ZFS for _building_ containers.

There's also some additional options in the `storage.conf` if you follow through the chain of links from the blogpost used in your example back to [the Proxmox forum thread that I think they got the info from](https://forum.proxmox.com/threads/podman-on-lxc-with-zfs-backed-volume-and-overlay.138722/):
```ini
[storage]
driver = "overlay"
runroot = "/run/containers/storage"
graphroot = "/var/lib/containers/storage"

[storage.options]
pull_options = {enable_partial_images = "false", use_hard_links = "false", ostree_repos=""}
mount_program = "/usr/local/bin/overlayzfsmount"

[storage.options.overlay]
mountopt = "nodev"
```

I had hoped those additional options would solve my issue, but it seems that sometimes when building containers, particularly when using the `COPY` operation to add files into a layer, I get a mount error:
```
wrong fs type, bad option, bad superblock on overlay, missing codepage or helper program, or other error.
       dmesg(1) may have more information after failed mount system call.
: exit status 32
```

There's nothing super helpful in `dmesg` on the host:
```
[Sun Sep  8 18:18:35 2024] overlayfs: fs on '/var/lib/containers/storage/overlay/l/EA3DCV4LX36CUSNQ35R6O6N64G' does not support file handles, falling back to xino=off.
[Sun Sep  8 18:18:35 2024] overlayfs: lowerdir is in-use as upperdir/workdir of another mount, accessing files from both mounts will result in undefined behavior.
[Sun Sep  8 18:19:15 2024] overlayfs: fs on '/var/lib/containers/storage/overlay/l/EA3DCV4LX36CUSNQ35R6O6N64G' does not support file handles, falling back to xino=off.
[Sun Sep  8 18:19:15 2024] overlayfs: fs on '/var/lib/containers/storage/overlay/l/5OMP6TYVQGBGLBOOYTCOMCE5TB' does not support file handles, falling back to xino=off.
[Sun Sep  8 18:19:15 2024] overlayfs: fs on '/var/lib/containers/storage/overlay/l/MDDQLWHHJECVGG3EBD6VAHZ35P' does not support file handles, falling back to xino=off.
```
